common/src/main/java/com/wynntils/mc/mixin/ScreenMixin.java

```diff
@@ -54,10 +59,39 @@ private void initPost(Minecraft client, int width, int height, CallbackInfo info
         EventFactory.onScreenCreated(screen, this::addRenderableWidget);
     }
 
+    @Redirect(
+            method = "renderTooltip(Lcom/mojang/blaze3d/vertex/PoseStack;Lnet/minecraft/world/item/ItemStack;II)V",
+            at =
+                    @At(
+                            value = "INVOKE",
+                            target =
+                                    "Lnet/minecraft/client/gui/screens/Screen;renderTooltip(Lcom/mojang/blaze3d/vertex/PoseStack;Ljava/util/List;Ljava/util/Optional;II)V"))
+    private void renderTooltipPre(
+            Screen instance,
+            PoseStack poseStack,
+            List<Component> tooltips,
+            Optional<TooltipComponent> visualTooltipComponent,
+            int mouseX,
+            int mouseY,
+            PoseStack poseStack2,
+            ItemStack itemStack,
+            int mouseX2,
+            int mouseY2) {
+        ItemTooltipRenderEvent e = EventFactory.onItemTooltipRenderPre(poseStack, itemStack, mouseX, mouseY);
+        if (e.isCanceled()) return;
+
+        instance.renderTooltip(
```

Comment:
@P0keDev I'm a bit nervous about this. Even if the TooltipScaleFeature is turned off, this will make the actual vanilla code behave differently, since you are not just tunneling the actual argument you received on to instance.renderTooltip. Is the idea that the event should be able to not only cancel the rendering, but to switch out the stack?

Is there some way we could do like "if the event was not sent, or not modified, call renderTooltip() exactly like in vanilla"?

Reply:
(In general, mixins apart from the trivial "inject which posts an event", should be kept to a minimum, and scrutinized carefully)

Reply:
if the event is unmodified it will execute the exact same code as vanilla, though - just in a slightly roundabout way. the same variables get passed to the same render function

Reply:
Maybe for the current version, but this risk breaking on each Minecraft update, in a hard-to-track way. I don't want to veto this code, but I'd like to get a shared understanding that this, and similar mixins, are high-risk.

Reply:
`ItemTooltipRenderEvent.Pre` isn't even canceled though (or at least I can't find where it is)?  I think we can rework this to be an inject and make it cancelable, that way it is less messy.

Reply:
What do you think about #83 @magicus? @P0keDev Can you confirm that the behavior is the same?

Reply:
> Maybe for the current version, but this risk breaking on each Minecraft update, in a hard-to-track way. I don't want to veto this code, but I'd like to get a shared understanding that this, and similar mixins, are high-risk.

it's certainly possible, but i think it's very unlikely for the vanilla code to change in a way that would break the mixin without also causing a compile error or mixin target failure

Reply:
Sure. I just want us to keep a high bar for adding non-inject mixins. I also believe they have a higher chance of interopability problems with other mods. I'm not saying we should ban them. They are like a sharp knife, a useful tool but you can cut yourself badly if you're not careful. And you should not use one when a dinner knife will do.